src/twisted/internet/test/test_iocp.py

```diff
@@ -210,3 +226,39 @@ def cb(self, rc, bytes, evt):
         self.assertEqual(fd.counter, EVENTS_PER_LOOP)
         ir.doIteration(0)
         self.assertEqual(fd.counter, EVENTS_PER_LOOP + 1)
+
+    def test_doubleHalfClose(self):
+        """
+        If one side half-closes its connection, and then the other side of the
+        connection calls C{loseWriteConnection}, and then C{loseConnection} in
+        {writeConnectionLost}, the connection is closed correctly.
+
+        This covers a code branch for ticket #9553.
+        """
+
+        @implementer(IHalfCloseableProtocol)
+        class ListenerProtocol(ConnectableProtocol):
+            def readConnectionLost(self):
+                # Force the existence of a writer in an attempt to cover #9553.
+                self.transport.startWriting()
+                self.transport.loseWriteConnection()
+
+            def writeConnectionLost(self):
+                self.transport.loseConnection()
+
+        class Client(ConnectableProtocol):
+            def connectionMade(self):
+                self.transport.loseConnection()
+
+        # If test fails, reactor won't stop and we'll hit timeout:
+        server = ListenerProtocol()
+        client = Client()
+        reactor = self.buildReactor()
+        reactor._handleSignals()
+        # We might have the waker readers.
+        initial_handles = reactor.handles.copy()
+
+        runProtocolsWithReactorInstance(reactor, self, server, client, TCPCreator())
```

Comment:
Rather than writing a whole new duplicated function which requires a matching reactor/reactorBuilder, it looks like what you want here is the ability to customize the reactor after it's built and before it's run.  Could you modify `runProtocolsWithReactor` to optionally take a callable of a function to run with the built reactor?  I feel that in addition to avoiding a ton of docstring duplication, having a callback would also make it clear what the "customize the reactor" bits are in the given callback and make it easier to see if there are other ways to unify those too.